### PR TITLE
Use alternate/faster linker if available

### PR DESCRIPTION
**Pull request by akohlmey**
**Summary**

Using either the gold or lld linker can drastically reduce the time to link executables. This is particularly noticeable when testing is enabled and not just the LAMMPS executable needs to be linked.

This is now a selectable (advanced) option and defaults to `lld`, then `gold`, then system default when not chosen explicitly.
See added documentation for more info.

**Author(s)**

Axel Kohlmeyer, Temple U

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No known issues. 

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the CMake based build system

### CODE REVIEW

File: cmake/Modules/Testing.cmake [hunk 1]
```diff
@@ -16,6 +16,18 @@ if(ENABLE_TESTING)
...
   set(MEMORYCHECK_COMMAND_OPTIONS "${VALGRIND_DEFAULT_OPTIONS}" CACHE STRING "Memory Check Command Options")
 
+  # check if a faster linker is available
+  include(CheckCXXCompilerFlag)
+  check_cxx_compiler_flag(-fuse-ld=lld HAVE_LLD_LINKER)
+  check_cxx_compiler_flag(-fuse-ld=gold HAVE_GOLD_LINKER)
+  if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.13)
+    if(HAVE_LLD_LINKER)
+      target_link_options(lammps PUBLIC -fuse-ld=lld)
+    elseif(HAVE_OLD_LINKER)
```
Comment (junghans) [inline cmake/Modules/Testing.cmake:26]: Typo? `HAVE_OLD_LINKER`-> `HAVE_GOLD_LINKER` ?

File: cmake/Modules/Testing.cmake [hunk 2]
```diff
@@ -16,6 +16,18 @@ if(ENABLE_TESTING)
   set(MEMORYCHECK_COMMAND "${VALGRIND_BINARY}" CACHE FILEPATH "Memory Check Command")
   set(MEMORYCHECK_COMMAND_OPTIONS "${VALGRIND_DEFAULT_OPTIONS}" CACHE STRING "Memory Check Command Options")
 
+  # check if a faster linker is available
+  include(CheckCXXCompilerFlag)
+  check_cxx_compiler_flag(-fuse-ld=lld HAVE_LLD_LINKER)
+  check_cxx_compiler_flag(-fuse-ld=gold HAVE_GOLD_LINKER)
+  if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.13)
```
Comment (junghans) [inline cmake/Modules/Testing.cmake:23]: If it is only for CMake>=3.13, this `if` should go on the outside. And maybe only do it for `ENABLE_TESTING=ON`.

File: cmake/Modules/Testing.cmake (hunk 1 above)
Comment (akohlmey) [reply cmake/Modules/Testing.cmake:26]: yes

File: cmake/Modules/Testing.cmake (hunk 2 above)
Comment (akohlmey) [reply cmake/Modules/Testing.cmake:23]: Yes, the if can be moved. But on the other hand, I submitted the PR to get some comments from @rbberger on whether he considers this kind of change useful for integration testing and in general. Then support for CMake 3.10-3.12 could be added as an else branch and the whole block moved to the toplevel CMakeList.txt file.

The entire file is already bracketed by a check of `ENABLE_TESTING`. That is why I put the check here in the first place.

**Comment by akohlmey** [comment]
@rbberger this is now complete as discussed for the "testing only" version. If you want to make changes to better utilize this from integration testing, grab the PR and feel free to shuffle things around. Otherwise, please approve.

**Pull request merged**